src/coreclr/jit/importer.cpp

```diff
@@ -8219,6 +8222,13 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
         szCanTailCallFailReason = "Caller is varargs";
     }
 #endif // FEATURE_FIXED_OUT_ARGS
+#if UNIX_X86_ABI
+    else
+    {
+        canTailCall             = false;
+        szCanTailCallFailReason = "Platform is Unix x86";
+    }
+#endif // UNIX_X86_ABI
```

Comment:
What problem did you run into with Unix x86 tailcalls? This is probably not right as it will deny even helper-based tailcalls that should just work on x86.
I think instead you should reject using the Windows x86 tailcall helper unconditionally in this function:
https://github.com/dotnet/runtime/blob/a579575853cea23aa1b6b6b8d3ef978aa9321e16/src/coreclr/jit/morph.cpp#L18172-L18191

Reply:
It is a fix of about 130 failing tests with message example:
```
FAILED   - [   2s]JIT/Directed/pinvoke/tail/tail.sh

               BEGIN EXECUTION
               /home/tmustafin/dotnet/x86_release_test/coreroot//corerun tail.dll ''
               Unhandled exception. System.InvalidProgramException: Common Language Runtime detected an invalid program.
                  at JitTest.Test.__DestroyMenu(UIntPtr hMenu)
                  at JitTest.Test.Main() in /home/tmustafin/dotnet/runtime/src/tests/JIT/Directed/pinvoke/tail.il:line 105
               ./tail.sh: line 375: 30509 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED

```
However, fix in `fgCanTailCallViaJitHelper` works better for couple of tests. Thanks!

Reply:
That's weird. Is it possible for you to retrieve a JIT dump from one of the failing tail call test cases?

Reply:
It is an example dump: https://gist.github.com/t-mustafin/2a60e785b148e7f496347979ea2bc26e.
Collected by:
```
DOTNET_JitDump="__DestroyMenu" ./corerun ~/dotnet/rebased2_x86_release_test/JIT/Directed/pinvoke/tail/tail.dll > ~/tasks/88_x86_helloworld/test_phase/tailcall_feedback/tail.dump
Unhandled exception. System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at JitTest.Test.__DestroyMenu(UIntPtr hMenu)
   at JitTest.Test.Main() in /home/tmustafin/dotnet/runtime/src/tests/JIT/Directed/pinvoke/tail.il:line 105
Aborted (core dumped)
```

Reply:
The dump seems to suggest that `fgCanTailCallViaJitHelper` returned true since it is still being morphed as a tail call via JIT helper:
```scala
fgMorphTailCallViaJitHelper (before):
               [000005] --CXG-------              *  CALL ind  void  
               [000000] ------------ arg0         +--*  LCL_VAR   int    V00 arg0         
               [000004] ------------ calli tgt    \--*  LCL_VAR   int    V01 tmp0         
fgMorphTailCallViaJitHelper (after):
               [000005] --CXG-------              *  CALL ind  void  
               [000000] ------------ arg0         +--*  LCL_VAR   int    V00 arg0         
               [000007] ------------ arg1         +--*  CNS_INT   int    0
               [000008] ------------ arg2         +--*  CNS_INT   int    9
               [000009] ------------ arg3         +--*  CNS_INT   int    8
               [000010] ------------ arg4         +--*  CNS_INT   int    7
               [000004] ------------ calli tgt    \--*  LCL_VAR   int    V01 tmp0         
```
Are you sure this is with your newest changes?

Reply:
It was on state before any tail call patch. With newest changes no tail call problem occures.

> However, fix in `fgCanTailCallViaJitHelper` works better for couple of tests. Thanks!
I mean that patch in `fgCanTailCallViaJitHelper` fixes all test which was fixed by previous patch + 2 more.

Reply:
Ah, my misunderstanding. Great to hear that it works.